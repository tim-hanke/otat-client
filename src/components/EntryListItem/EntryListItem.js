import React, { Component } from "react";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt } from "@fortawesome/free-regular-svg-icons";
import EntryApiService from "../../services/entry-api-service";
import "./EntryListItem.css";
import ArticleListContext from "../../contexts/ArticleListContext";

export default class EntryListItem extends Component {
  static contextType = ArticleListContext;

  state = { error: null };

  handleClickDelete = async () => {
    const { id } = this.props.article;

    try {
      await EntryApiService.deleteUserArticle(id);
      EntryApiService.getEntries()
        .then(this.context.setEntryList)
        .catch(this.context.setError);
      this.setState({ error: null });
    } catch (res) {
      this.setState({ error: res.error });
    }
  };

  render() {
    const { article } = this.props;
    const { error } = this.state;

    return (
      <div className="EntryListItem__container">
        <Link
          to={{
            pathname: article.url,
          }}
          target="_blank"
          className="EntryListItem"
          rel="noopener noreferrer"
        >
          <div
            className="EntryListItem__image"
            style={{ backgroundImage: `url(${article.image})` }}
          />

          <div className="EntryListItem__details">
            <div className="EntryListItem__text">
              <h2 className="EntryListItem__title">{article.title}</h2>
              <p className="EntryListItem__description">
                {truncate(article.description)}
              </p>
            </div>
          </div>
          {error && (
            <div role="alert">
              <p>{error}</p>
            </div>
          )}
        </Link>
        <button
          className="EntryListItem__delete jiggly"
          onClick={this.handleClickDelete}
          aria-label="delete button"
          title="delete button"
        >
          <FontAwesomeIcon className="logo" icon={faTrashAlt} />
        </button>
      </div>
    );
  }
}

function truncate(text) {
  const words = text.split(" ");

  if (words.length > 40) {
    return words.slice(0, 40).join(" ") + " ...";
  }

  return text;
}
