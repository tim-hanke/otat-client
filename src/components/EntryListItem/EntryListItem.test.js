import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import EntryListItem from "./EntryListItem";

describe.only("EntryListItem component", () => {
  const testEntry = {
    id: 1,
    date_created: "2001-01-01",
    text:
      "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Natus consequuntur deserunt commodi, nobis qui inventore corrupti iusto aliquid debitis unde non.Adipisci, pariatur.Molestiae, libero esse hic adipisci autem neque ?",
    user_id: 1,
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <EntryListItem entry={testEntry} />
      </MemoryRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <EntryListItem entry={testEntry} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
