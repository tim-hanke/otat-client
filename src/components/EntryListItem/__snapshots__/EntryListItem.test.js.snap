// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryListItem component renders the UI as expected 1`] = `
<div
  className="EntryListItem__container"
>
  <div
    className="EntryListItem"
  >
    <div
      className="EntryListItem__details"
    >
      <div
        className="EntryListItem__text"
      >
        <h2
          className="EntryListItem__title"
        >
          2001-01-01
        </h2>
        <p
          className="EntryListItem__description"
        >
          Lorem ipsum dolor sit amet, consectetur adipisicing elit. Natus consequuntur deserunt commodi, nobis qui inventore corrupti iusto aliquid debitis unde non.Adipisci, pariatur.Molestiae, libero esse hic adipisci autem neque ?
        </p>
      </div>
    </div>
  </div>
  <button
    aria-label="delete button"
    className="EntryListItem__delete jiggly"
    onClick={[Function]}
    title="delete button"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-trash-alt fa-w-14 logo"
      data-icon="trash-alt"
      data-prefix="far"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M268 416h24a12 12 0 0 0 12-12V188a12 12 0 0 0-12-12h-24a12 12 0 0 0-12 12v216a12 12 0 0 0 12 12zM432 80h-82.41l-34-56.7A48 48 0 0 0 274.41 0H173.59a48 48 0 0 0-41.16 23.3L98.41 80H16A16 16 0 0 0 0 96v16a16 16 0 0 0 16 16h16v336a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128h16a16 16 0 0 0 16-16V96a16 16 0 0 0-16-16zM171.84 50.91A6 6 0 0 1 177 48h94a6 6 0 0 1 5.15 2.91L293.61 80H154.39zM368 464H80V128h288zm-212-48h24a12 12 0 0 0 12-12V188a12 12 0 0 0-12-12h-24a12 12 0 0 0-12 12v216a12 12 0 0 0 12 12z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</div>
`;
